feat: add kyverno authz server blog post (#15867)
istio/istio.io: content/en/blog/2024/authz-policy-with-kyverno/index.md

---

> title: Policy based authorization - Using Kyverno Authz Server
> description: Delegate authorization decisions logic to Kyverno Authz Server, leveraging Kyverno authorization policies based on CEL bringing safety, flexibility, and unprecedented performance.

Reviewer: ```suggestion
title: title: Policy based authorization using Kyverno
description: Delegate Layer 7 authorization decision logic using Kyverno's authz server, leveraging policies based on CEL.
```

It's unclear from the docs & GitHub page for your Envoy plugin if you want "Kyverno Authz Server" to be a Brand Name throughout.  If you don't, I'd recommend you change to "Kyverno authz server" throughout the doc.

Author: My feeling is that "Kyverno Authz Server" is better to avoid confusion with Kyverno (the policy engine).

publishdate: 2024-11-20
attribution: "Charles-Edouard Brétéché (Nirmata)"
keywords: [istio,kyverno,policy,platform,authorization]
---

> A couple of weeks ago, the [Can Your Platform Do Policy? Accelerate Teams With Platform L7 Policy Functionality](../l7-policy-with-opa) blog post was published.
> 
> This blog post is a follow-up article, demonstrating the same capacity with the Kyverno Authz Server instead of OPA.
> 
> Today, we're going to dive into how Istio and the Kyverno Authz Server can be used together to enforce Layer 7 policies in your platform.
> We'll show you how to get started with a simple example.
> You will come to see how this combination is a solid option to deliver policy quickly and transparently to application team everywhere in the business, while also providing the data the security teams need for audit and compliance.

Reviewer: ```suggestion
Istio supports integration with many different projects.  The Istio blog recently featured a post on [L7 policy functionality with OpenPolicyAgent](../l7-policy-with-opa). Kyverno is a similar project, and today we will dive how Istio and the Kyverno Authz Server can be used together to enforce Layer 7 policies in your platform.
We will show you how to get started with a simple example.
You will come to see how this combination is a solid option to deliver policy quickly and transparently to application team everywhere in the business, while also providing the data the security teams need for audit and compliance.
```
perhaps

## Try it out

When integrated with Istio, the Kyverno Authz Server can be used to enforce fine-grained access control policies for microservices.

This guide shows how to enforce access control policies for a simple microservices application.

### Prerequisites

- A Kubernetes cluster with Istio installed.
- The `istioctl` command-line tool installed.

Install Istio and configure your [mesh options](/docs/reference/config/istio.mesh.v1alpha1/) to enable Kyverno:

{{< text bash >}}
$ istioctl install -y -f - <<EOF
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
spec:
  meshConfig:
    accessLogFile: /dev/stdout
    accessLogFormat: |
      [KYVERNO DEMO] my-new-dynamic-metadata: '%DYNAMIC_METADATA(envoy.filters.http.ext_authz)%'
    extensionProviders:

>     - name: kyverno-authz-server.local

Reviewer: I find it confusing that this has the appearance of a DNS name; would just `kyverno-auth-server` make more sense throughout?

      envoyExtAuthzGrpc:
        service: kyverno-authz-server.kyverno.svc.cluster.local
        port: '9081'
EOF
{{< /text >}}

Notice that in the configuration, we define an `extensionProviders` section that points to the Kyverno Authz Server installation:

{{< text yaml >}}
[...]
    extensionProviders:
    - name: kyverno-authz-server.local
      envoyExtAuthzGrpc:
        service: kyverno-authz-server.kyverno.svc.cluster.local
        port: '9081'
[...]
{{< /text >}}

#### Deploy the Kyverno Authz Server

The Kyverno Authz Server is a GRPC server capable of processing Envoy External Authorization requests.

It is configurable using Kyverno `AuthorizationPolicy` resources, either stored in-cluster or provided externally.

> {{< text bash >}}
> $ kubectl create ns kyverno
> {{< /text >}}
> 
> {{< text bash >}}
> $ kubectl label namespace kyverno istio-injection=enabled
> {{< /text >}}
> 
> {{< text bash >}}
> $ helm install kyverno-authz-server --namespace kyverno --wait --repo https://kyverno.github.io/kyverno-envoy-plugin kyverno-authz-server
> {{< /text >}}

Reviewer: ```suggestion
{{< text bash >}}
$ kubectl create ns kyverno
$ kubectl label namespace kyverno istio-injection=enabled
$ helm install kyverno-authz-server --namespace kyverno --wait --repo https://kyverno.github.io/kyverno-envoy-plugin kyverno-authz-server
{{< /text >}}
```

#### Deploy the sample application

> Httpbin is a well-known application that can be used to test HTTP requests and helps to show quickly how we can play with the request and response attributes.

Reviewer: ```suggestion
httpbin is a well-known application that can be used to test HTTP requests and helps to show quickly how we can play with the request and response attributes.
```

should have caught this in the OPA blog too

> {{< text bash >}}
> $ kubectl create ns my-app
> {{< /text >}}
> 
> {{< text bash >}}
> $ kubectl label namespace my-app istio-injection=enabled
> {{< /text >}}
> 
> {{< text bash >}}
> $ kubectl apply -f {{< github_file >}}/samples/httpbin/httpbin.yaml -n my-app
> {{< /text >}}

Reviewer: ```suggestion
{{< text bash >}}
$ kubectl create ns my-app
$ kubectl label namespace my-app istio-injection=enabled
$ kubectl apply -f {{< github_file >}}/samples/httpbin/httpbin.yaml -n my-app
{{< /text >}}
```

#### Deploy an Istio AuthorizationPolicy

An `AuthorizationPolicy` defines the services that will be protected by the Kyverno Authz Server.

{{< text bash >}}
$ kubectl apply -f - <<EOF
apiVersion: security.istio.io/v1
kind: AuthorizationPolicy
metadata:
  name: my-kyverno-authz

>   namespace: istio-system # This enforce the policy on all the mesh being istio-system the mesh config namespace

Reviewer: ```suggestion
  namespace: istio-system # This enforce the policy on all the mesh, istio-system being the mesh root namespace
```

spec:
  selector:
    matchLabels:
      ext-authz: enabled
  action: CUSTOM
  provider:
    name: kyverno-authz-server.local
  rules: [{}] # Empty rules, it will apply to selectors with ext-authz: enabled label
EOF
{{< /text >}}

Notice that in this resource, we define the Kyverno Authz Server `extensionProvider` you set in the Istio configuration:

{{< text yaml >}}
[...]
  provider:
    name: kyverno-authz-server.local
[...]
{{< /text >}}

#### Label the app to enforce the policy

> 

Reviewer: put a small description here of what you're doing?  Would you not normally just use `kubectl label` in this case?

Author: > Would you not normally just use kubectl label in this case?

Not really because we are adding the label to the pod template, not to the deployment.

{{< image link="./filters-chain.svg" alt="Filters" >}}

Reviewing [Envoy's Authorization service documentation](https://www.envoyproxy.io/docs/envoy/latest/api-v3/service/auth/v3/external_auth.proto), you can see that the message has these attributes:

- Ok response

    {{< text json >}}
    {
      "status": {...},
      "ok_response": {
        "headers": [],
        "headers_to_remove": [],
        "response_headers_to_add": [],
        "query_parameters_to_set": [],
        "query_parameters_to_remove": []
      },
      "dynamic_metadata": {...}
    }
    {{< /text >}}

- Denied response

    {{< text json >}}
    {
      "status": {...},
      "denied_response": {
        "status": {...},
        "headers": [],
        "body": "..."
      },
      "dynamic_metadata": {...}
    }
    {{< /text >}}

This means that based on the response from the authz server, Envoy can add or remove headers, query parameters, and even change the response body.

> The Kyverno Authz Server can do this as well, as documented in the [Kyverno Authz Server documentation](https://kyverno.github.io/kyverno-envoy-plugin).

Reviewer: You _are_ the authz server?  Does "we can do this as well" make sense?

## Testing

Let's test the simple usage (authorization) and then let's create a more advanced policy to show how we can use the Kyverno Authz Server to modify the request and response.

Deploy an app to run curl commands to the httpbin sample application:

{{< text bash >}}

> $ kubectl -n my-app run --image=curlimages/curl curl -- /bin/sleep 100d

Reviewer: You could use 

```
$ kubectl apply -f {{< github_file >}}/samples/curl/curl.yaml
```

and then 

```
$ kubectl exec deploy/curl -- curl -s "http://productpage:9080/productpage"
```

as your command
